gdscript/sidebar.gd

```diff
@@ -1032,95 +1044,163 @@ enum HistoryListPopupItem {
+	if button_idx == MOUSE_BUTTON_RIGHT:
+		# if the selected item is disabled, do not.
+		if history_tree.get_selected().get_metadata(1) == false: return
+		show_contextmenu(history_tree.get_selected().get_metadata(0))
+
+func show_contextmenu(item_hash):
+		context_menu_hash = item_hash
+		history_list_popup.position = DisplayServer.mouse_get_position()
+		history_list_popup.visible = true
+
+func _on_history_tree_button_clicked(item: TreeItem, _column : int, id: int, mouse_button_index: int) -> void:
+	if mouse_button_index != MOUSE_BUTTON_LEFT: return
+	
+	if id == 0:
+		var change_hash = item.get_metadata(0)
+		var history = GodotProject.get_changes()
+
+		history = history.filter(func (c): return c.hash == change_hash);
+		if history.is_empty():
+			print("Error: No matching change found.")
+			return;
+
+		var change = history[0]
+		var merged_branch = GodotProject.get_branch_by_id(change.merge_metadata.merged_branch_id)
+		checkout_branch(merged_branch.id)
+	elif id == 1:
+		show_contextmenu(item.get_metadata(0))
+
+func _on_history_list_item_selected() -> void:
+	var selected_item = history_tree.get_selected()
+	if selected_item == null:
+		history_saved_selection = null
 		return
-	else:
-		print("Right clicked index is not disabled: ", index)
-	history_list_popup.position = DisplayServer.mouse_get_position()
-	history_list_popup.visible = true
 
+	# update the saved selection
+	var change_hash = selected_item.get_metadata(0)
+	history_saved_selection = change_hash
 
-func _on_history_list_item_selected(index: int, _at_position: Vector2, button_idx: int) -> void:
-	if button_idx == MOUSE_BUTTON_RIGHT:
-		return _on_item_right_clicked(index, _at_position, button_idx)
+	update_diff()
 
+func _on_clear_diff_button_pressed():
+	_on_history_tree_empty_clicked(null, 0)
 
-	# check if the mouse position is inside the icon rect
-	if _at_position.x <= 64:
-		return _on_item_right_clicked(index, _at_position, button_idx)
+func _on_history_tree_empty_clicked(_vec2, _idx):
+	history_saved_selection = null
+	history_tree.deselect_all()
+	update_diff()
 
-	var history_size = history_list.get_item_count()
+# read the selection from the tree, and update the diff visualization accordingly.
+func update_diff():
+	var selected_item = history_tree.get_selected()
 	var checked_out_branch = GodotProject.get_checked_out_branch()
-	if (index >= history_size or (checked_out_branch.is_main and index >= history_size - 3)):
-		# the first three commits on main are initial checkin, so we don't show them
+
+	# check to see if we generate a selection diff, or a default diff
+	if (selected_item == null
+			or checked_out_branch.is_merge_preview
+			or checked_out_branch.is_revert_preview
+			# the first two commits on main are initial checkin, so we don't show a diff for them
+			or checked_out_branch.is_main
+				and selected_item.get_index() >= history_item_count - 2):
+		update_diff_default(checked_out_branch, GodotProject.get_changes().size())
```

Comment:
`GodotProject.get_changes()` is an expensive operation, so we don't want to do it unless we have to. Try to refactor this to minimize the calls to it